README
NAME
    Path::IsDev - Determine if a given Path resembles a development source
    tree

VERSION
    version 1.001003

SYNOPSIS
        use Path::IsDev qw(is_dev);

        if( is_dev('/some/path') ) {
            ...
        } else {
            ...
        }

DESCRIPTION
    This module is more or less a bunch of heuristics for determining if a
    given path is a development tree root of some kind.

    This has many useful applications, notably ones that require behaviours
    for "installed" modules to be different to those that are still "in
    development"

FUNCTIONS
  debug
    Debug callback.

    To enable debugging:

        export PATH_ISDEV_DEBUG=1

  "is_dev"
    Using an "import"'ed "is_dev":

        if( is_dev( $path ) ) {

        }

    Though the actual heuristics used will be based on how "import" was
    called.

    Additionally, you can call

        Path::IsDev::is_dev

    without "import"ing anything, and it will behave exactly the same as if
    you'd imported it using

        use Path::IsDev qw( is_dev );

    That is, no "set" specification is applicable, so you'll only get the
    "default".

UNDERSTANDING AND DEBUGGING THIS MODULE
    Understanding how this module works, is critical to understand where you
    can use it, and the consequences of using it.

    This module operates on a very simplistic level, and its easy for
    false-positives to occur.

    There are two types of Heuristics, Postive/Confirming Heuristics, and
    Negative/Disconfirming Heuristics.

    Positive Heuristics and Negative Heuristics are based solely on the
    presence of specific marker files in a directory, or special marker
    directories.

    For instance, the files "META.yml", "Makefile.PL", and "Build.PL" are
    all Positive Heuristic markers, because their presence often indicates a
    "root" of a development tree.

    And for instance, the directories "t/", "xt/" and ".git/" are also
    Positive Heuristic markers, because these structures are common in
    "perl" development trees, and uncommon in install trees.

    However, these markers sometimes go wrong, for instance, consider you
    have a "local::lib" or "perlbrew" install in $HOME

        $HOME/
        $HOME/lib/
        $HOME/perl5/perls/perl-5.19.3/lib/site_perl/

    Etc.

    Under normal circumstances, neither $HOME nor those 3 paths are
    considered "dev".

    However, all it takes to cause a false positive, is for somebody to
    install a "t" or "xt" directory, or a marker file in one of the above
    directories for "path_isdev($dir)" to return true.

    This may not be a problem, at least, until you use "Path::FindDev" which
    combines "Path::IsDev" with recursive up-level traversal.

        $HOME/
        $HOME/lib/
        $HOME/perl5/perls/perl-5.19.3/lib/site_perl/

        find_dev('$HOME/perl5/perls/perl-5.19.3/lib/site_perl/') # returns false, because it is not inside a dev directory

        mkdir $HOME/t

        find_dev('$HOME/perl5/perls/perl-5.19.3/lib/site_perl/') # returns $HOME, because $HOME/t exists.

    And it is this kind of problem that usually catches people off guard.

        PATH_ISDEV_DEBUG=1 \
            perl -Ilib -MPath::FindDev=find_dev \
            -E "say find_dev(q{/home/kent/perl5/perlbrew/perls/perl-5.19.3/lib/site_perl})"

        ...
        [Path::IsDev=0] + ::Tool::Dzil => 0 : dist.ini does not exist
        [Path::IsDev=0] + ::Tool::MakeMaker => 0 : Makefile.PL does not exist
        [Path::IsDev=0] + ::Tool::ModuleBuild => 0 : Build.PL does not exist
        [Path::IsDev=0] + ::META => 0 : META.json does not exist
        [Path::IsDev=0] + ::META => 1 : META.yml exists
        [Path::IsDev=0] + ::META => 1 : /home/kent/perl5/META.yml is a file
        [Path::IsDev=0] + ::META matched path /home/kent/perl5
        /home/kent/perl5

    Whoops!.

        [Path::IsDev=0] + ::META => 1 : META.yml exists
        [Path::IsDev=0] + ::META => 1 : /home/kent/perl5/META.yml is a file

    No wonder!

        rm /home/kent/perl5/META.yml

        PATH_ISDEV_DEBUG=1 \
            perl -Ilib -MPath::FindDev=find_dev \
            -E "say find_dev(q{/home/kent/perl5/perlbrew/perls/perl-5.19.3/lib/site_perl})"

        ...
        [Path::IsDev=0] Matching /home/kent/perl5
        ...
        [Path::IsDev=0] + ::TestDir => 0 : xt does not exist
        [Path::IsDev=0] + ::TestDir => 1 : t exists
        [Path::IsDev=0] + ::TestDir => 1 : /home/kent/perl5/t is a dir
        [Path::IsDev=0] + ::TestDir matched path /home/kent/perl5
        /home/kent/perl5

    Double whoops!

        [Path::IsDev=0] + ::TestDir => 1 : t exists
        [Path::IsDev=0] + ::TestDir => 1 : /home/kent/perl5/t is a dir

    And you could keep doing that until you rule out all the bad heuristics
    in your tree.

    Or, you could use a negative heuristic.

        touch /home/kent/perl5/.path_isdev_ignore

        PATH_ISDEV_DEBUG=1 \
            perl -Ilib -MPath::FindDev=find_dev \
            -E "say find_dev(q{/home/kent/perl5/perlbrew/perls/perl-5.19.3/lib/site_perl})"
        ...
        [Path::IsDev=0] Matching /home/kent/perl5
        [Path::IsDev=0] - ::IsDev::IgnoreFile => 1 : .path_isdev_ignore exists
        [Path::IsDev=0] - ::IsDev::IgnoreFile => 1 : /home/kent/perl5/.path_isdev_ignore is a file
        [Path::IsDev=0] - ::IsDev::IgnoreFile excludes path /home/kent/perl5
        [Path::IsDev=0] no match found
        ...
        [Path::IsDev=0] Matching /
        ...
        [Path::IsDev=0] no match found

    Success!

        [Path::IsDev=0] - ::IsDev::IgnoreFile => 1 : .path_isdev_ignore exists
        [Path::IsDev=0] - ::IsDev::IgnoreFile => 1 : /home/kent/perl5/.path_isdev_ignore is a file

HEURISTICS
  Negative Heuristics bundled with this distribution
    Just remember, a Negative Heuristic excludes the path it is associated
    with

    *   "IsDev::IgnoreFile" - ".path_isdev_ignore"

  Positive Heuristics bundled with this distribution
    *   "Changelog" - Files matching "Changes", "Changelog", and similar,
        case insensitive, extensions optional.

    *   "DevDirMarker" - explicit ".devdir" file to indicate a project root.

    *   "META" - "META.yml"/"META.json"

    *   "MYMETA" - "MYMETA.yml"/"MYMETA.json"

    *   "Makefile" - Any "Makefile" format documented supported by GNU Make

    *   "TestDir" - A directory called either "t/" or "xt/"

    *   "Tool::DZil" - A "dist.ini" file

    *   "Tool::MakeMaker" - A "Makefile.PL" file

    *   "Tool::ModuleBuild" - A "Build.PL" file

    *   "VCS::Git" - A ".git" directory

HEURISTIC SETS
  Heuristic Sets Bundled with this distribution
    *   "Basic" - The basic heuristic set that contains most, if not all
        heuristics.

ADVANCED USAGE
  Custom Sets
    "Path::IsDev" has a system of "sets" of Heuristics, in order to allow
    for pluggable and flexible heuristic types.

    Though, for the vast majority of cases, this is not required.

        use Path::IsDev is_dev => { set => 'Basic' };
        use Path::IsDev is_dev => { set => 'SomeOtherSet' , -as => 'is_dev_other' };

  Overriding the default set
    If for whatever reason the "Basic" set is insufficient, or if it false
    positives on your system for some reason, the "default" set can be
    overridden.

        export PATH_ISDEV_DEFAULT_SET="SomeOtherSet"

        ...
        use Path::IsDev qw( is_dev );
        is_dev('/some/path') # uses SomeOtherSet

    Though this will only take priority in the event the set is not
    specified during "import"

    If this poses a security concern for the user, then this security hole
    can be eliminated by declaring the set you want in code:

        export PATH_ISDEV_DEFAULT_SET="SomeOtherSet"

        ...
        use Path::IsDev  is_dev => { set => 'Basic' };
        is_dev('/some/path') # uses Basic, regardless of ENV

SECURITY
    Its conceivable, than an evil user could construct an evil set,
    containing arbitrary and vulnerable code, and possibly stash that evil
    set in a poorly secured privileged users @INC

    And if they managed to achieve that, if they could poison the privileged
    users %ENV, they could trick the privileged user into executing
    arbitrary code.

    Though granted, if you can do either of those 2 things, you're probably
    security vulnerable anyway, and granted, if you could do either of those
    2 things you could do much more evil things by the following:

        export PERL5OPT="-MEvil::Module"

    So with that in understanding, saying this modules default utility is
    "insecure" is mostly a bogus argument.

    And to that effect, this module does nothing to "lock down" that
    mechanism, and this module encourages you to NOT force a set, unless you
    NEED to, and strongly suggests that forcing a set for the purpose of
    security will achieve no real improvement in security, while
    simultaneously reducing utility.

AUTHOR
    Kent Fredric <kentfredric@gmail.com>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2014 by Kent Fredric
    <kentfredric@gmail.com>.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Release history for Path-IsDev

1.001004 2017-03-09T20:40:16Z 38fe9b0

1.001003 2017-03-09T20:21:10Z 4dcd533
 - Bugfix: Avoid -Ddefault_inc_excludes_dot failures ( also makes tests *much* faster )

 [Dependencies::Stats]
 - Dependencies changed since 1.001002, see misc/*.deps* for details
 - configure: +1 (recommends: ↑1)
 - develop: +14 ↑4 -3 (recommends: +1, suggests: ↑2)
 - runtime: (recommends: +1)
 - test: +1 ↓1 (recommends: ↑1 ↓1)

1.001002 2014-08-16T04:40:51Z
 [00 Trivial]
 - CPANDAY!
 - no code changes
 - tests augmented
 - whitespace reduced

 [Dependencies::Stats]
 - Dependencies changed since 1.001001, see misc/*.deps* for details
 - configure: ↓1
 - develop: +1 (suggests: ↑2)
 - test: +1 -3 (recommends: ↑1 -1)

 [Metadata]
 - move to kentnl org

 [Misc]
 - Improve source side POD to reduce build side whitespace.

 [Tests]
 - update Test::ReportPrereqs

1.001001 2014-07-12T18:23:30Z
 [00 Trivial]
 - Misc packaging cleanups.

 [Dependencies::Stats]
 - Dependencies changed since 1.001000, see misc/*.deps* for details
 - configure: (recommends: ↑1)
 - develop: +3 -3 (suggests: +2 -1)
 - test: (recommends: ↑2)

1.001000 2014-03-01T13:22:00Z
 [00 Majorish]
 - Tooling switched to ExtUtils::MakeMaker

 [Bug fixes]
 - Needed newer Path::Tiny for ->realpath

 [Dependencies::Aggregate]
 - 5.008 - 48 → 43
 - 5.010 - 31 → 32
 - 5.019 - 17 → 16

 [Dependencies::Stats]
 - Dependencies changed since 1.000002, see misc/*.deps* for details
 - build: -1
 - configure: +1 -1 (recommends: +1)
 - develop: +7 ↑1 -9
 - runtime: +1 ↑2
 - test: +4 ↓1 (recommends: +4)

1.000002 2013-11-22T03:28:04Z
 [Bug Fixes]
 - Attempt to avoid failures on some Win32's where methods other than File::HomeDir->my_home don't work at all.

 [Dependencies::Stats]
 - Dependencies changed since 1.000001, see misc/*.deps* for details
 - develop: ↑1

 [Maintenance]
 - Split Changes into Changes.{deps,deps.all,deps.dev} for easier reading for maintainers.

1.000001 2013-11-21T11:49:20Z
 [Dependencies::Stats]
 - Dependencies changed since 1.000000, see misc/*.deps* for details
 - build: ↑1
 - configure: ↑1
 - develop: +42 ↑1
 - test: ↑1 -3

 [Documentation]
 - README.mkdn → MetaCPAN
 - Encoding → UTF-8

 [Maintenance]
 - Increase deps on Module::Build and Test::More
 - Removes a few deps no longer needed.

 [Packaging]
 - Use new fallback code in MB

 [Tests]
 - Replace Test::Compile with Test::Compile::PerFile

1.000000 2013-10-18T09:55:20Z
 [00 Supermajor Changes - Breaking Everything]
 - This release comes with many significant changes that are likely to break any closely built dependencies.
 - External export interface contains no visible changes, but underyling plumbing is under heavy construction.

 [01 Role Refactory]
 - Any existing code that relied on previous roles or base classes will be BROKEN by this release.
 - Roles again reimplemented and split to be more horizonal instead of vertically integrated.
 - This means all Heuristics need to provide a ->matches() or ->excludes() method now, and manually call their
   composite roles, and manually set $result_object->result(1) where relevant.

 [02 Base class removal]
 - Base classes that we're deprecated in 0.6.0 are gone now.

 [03 Version Schema Change]
 - Version scheme is now Y.nnnnnn . If this is a problem for you, this means you're a downstream who doesn't normalise
   versions yet. Please learn to normalise, perl upstream break your assumptions.
 - Versions will retain semantic difference, just there more "perl internals" friendly.

 [04 New Heuristic Match Types]
 - Now a Matcher that matches on full paths

 [05 New Negative Heuristics]
 - ::HomeDir - Excludes all the paths returned by File::HomeDir
 - ::PerlINC - Excludes all the paths in Config.pm

 [06 Updated Heuristic Sets]
 - ::Basic now includes ::HomeDir and ::PerlINC exclusions

 [07 Internals]
 - Requirement for matches to debug themselves greatly reduced.
 - Matches now only need to report context feedback via the result object
 - The result object intelligently executes relevant debugging intel

 [Dependencies]
 - Now depends on File::HomeDir
 - Perl 5.6 is now minimum ( was 5.10 )
 - Test::More increased.

 [Dependencies::Stats]
 - Dependencies changed since 0.6.0, see misc/*.deps* for details
 - develop: ↑2
 - runtime: +1 ↓1
 - test: ↑1

0.6.0 2013-10-05T17:56:06Z
 [00 Major Changes - Role Apocalypse]
 - All internals that could be rewritten to invoke roles instead of inheritance, have been done so.
 - Base classes for existing things still exist, and should still work, however, they're on borrowed time.

 [Dependencies]
 - Now requires Role::Tiny
 - No longer requires parent

 [Dependencies::Stats]
 - Dependencies changed since 0.5.0, see misc/*.deps* for details
 - runtime: +2 -1

0.5.0 2013-10-05T15:00:12Z
 [00 Major Changes - Result Object]
 - Internals now re-written slightly to pass around a result object that communicates state.
 - Everywhere $path was previously used, $result_object->path is now used instead.
 - $result_object->path guaranteed to exist and be a ref of some kind.
 - $result_object->path automatically upgraded to Path::Tiny if provided as a string
 - ->matches() and ->excludes() can now record results of heuristics by invoking $result_object->add_reason()
 - Path::IsDev::Object->_matches($path) now exists and returns a Result Object for $path
 - Dumping result object will give you as much info as turning debugging on would

0.4.0 2013-09-27T17:48:00Z
 [00 Major Changes - Negative Heuristics]
 - Now has a system to indicate negative heuristics, to mark  given directories as "not dev", which cause other postive
   heuristics to be ignored.
 - One basic negative heuristic added, 'IsDev::IgnoreFile'
 - Basic Negative heuristic comes standard with the basic set.

 [Dependencies::Stats]
 - Dependencies changed since 0.3.3, see misc/*.deps* for details
 - develop: +1 -1 (recommends: -1, suggests: +1 -1)

0.3.3 2013-09-26T07:09:14Z
 [Bugfix]
 - Use a Non-Win32-failing compile test

lib/Path/IsDev.pm
use 5.008;    # utf8
use strict;
use warnings;
use utf8;

package Path::IsDev;

our $VERSION = '1.001004';

# ABSTRACT: Determine if a given Path resembles a development source tree

our $AUTHORITY = 'cpan:KENTNL'; # AUTHORITY












use Sub::Exporter -setup => { exports => [ is_dev => \&_build_is_dev, ], };

our $ENV_KEY_DEBUG = 'PATH_ISDEV_DEBUG';

our $DEBUG = ( exists $ENV{$ENV_KEY_DEBUG} ? $ENV{$ENV_KEY_DEBUG} : undef );











sub debug {
  return unless $DEBUG;
  return *STDERR->printf( qq{[Path::IsDev] %s\n}, shift );
}

sub _build_is_dev {
  my ( undef, undef, $arg ) = @_;

  my $isdev_object;
  return sub {
    my ($path) = @_;
    $isdev_object ||= do {
      require Path::IsDev::Object;
      Path::IsDev::Object->new( %{ $arg || {} } );
    };
    return $isdev_object->matches($path);
  };
}
























*is_dev = _build_is_dev( 'Path::IsDev', 'is_dev', {} );

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Path::IsDev - Determine if a given Path resembles a development source tree

=head1 VERSION

version 1.001004

=head1 SYNOPSIS

    use Path::IsDev qw(is_dev);

    if( is_dev('/some/path') ) {
        ...
    } else {
        ...
    }

=head1 DESCRIPTION

This module is more or less a bunch of heuristics for determining if a given path
is a development tree root of some kind.

This has many useful applications, notably ones that require behaviours for "installed"
modules to be different to those that are still "in development"

=head1 FUNCTIONS

=head2 debug

Debug callback.

To enable debugging:

    export PATH_ISDEV_DEBUG=1

=head2 C<is_dev>

Using an C<import>'ed C<is_dev>:

    if( is_dev( $path ) ) {

    }

Though the actual heuristics used will be based on how C<import> was called.

Additionally, you can call

    Path::IsDev::is_dev

without C<import>ing anything, and it will behave exactly the same as if you'd imported
it using

    use Path::IsDev qw( is_dev );

That is, no C<set> specification is applicable, so you'll only get the "default".

=begin MetaPOD::JSON v1.1.0

{
    "namespace":"Path::IsDev",
    "interface":"exporter"
}


=end MetaPOD::JSON

=head1 UNDERSTANDING AND DEBUGGING THIS MODULE

Understanding how this module works, is critical to understand where you can use it, and the consequences of using it.

This module operates on a very simplistic level, and its easy for false-positives to occur.

There are two types of Heuristics, Postive/Confirming Heuristics, and Negative/Disconfirming Heuristics.

Positive Heuristics and Negative Heuristics are based solely on the presence of specific marker files in a directory, or special
marker directories.

For instance, the files C<META.yml>, C<Makefile.PL>, and C<Build.PL> are all B<Positive Heuristic> markers, because their
presence often indicates a "root" of a development tree.

And for instance, the directories C<t/>, C<xt/> and C<.git/> are also B<Positive Heuristic> markers, because these structures
are common in C<perl> development trees, and uncommon in install trees.

However, these markers sometimes go wrong, for instance, consider you have a C<local::lib> or C<perlbrew> install in C<$HOME>

    $HOME/
    $HOME/lib/
    $HOME/perl5/perls/perl-5.19.3/lib/site_perl/

Etc.

Under normal circumstances, neither C<$HOME> nor those 3 paths are considered C<dev>.

However, all it takes to cause a false positive, is for somebody to install a C<t> or C<xt> directory, or a marker file in one of
the above directories for C<path_isdev($dir)> to return true.

This may not be a problem, at least, until you use C<Path::FindDev> which combines C<Path::IsDev> with recursive up-level
traversal.

    $HOME/
    $HOME/lib/
    $HOME/perl5/perls/perl-5.19.3/lib/site_perl/

    find_dev('$HOME/perl5/perls/perl-5.19.3/lib/site_perl/') # returns false, because it is not inside a dev directory

    mkdir $HOME/t

    find_dev('$HOME/perl5/perls/perl-5.19.3/lib/site_perl/') # returns $HOME, because $HOME/t exists.

And it is this kind of problem that usually catches people off guard.

    PATH_ISDEV_DEBUG=1 \
        perl -Ilib -MPath::FindDev=find_dev \
        -E "say find_dev(q{/home/kent/perl5/perlbrew/perls/perl-5.19.3/lib/site_perl})"

    ...
    [Path::IsDev=0] + ::Tool::Dzil => 0 : dist.ini does not exist
    [Path::IsDev=0] + ::Tool::MakeMaker => 0 : Makefile.PL does not exist
    [Path::IsDev=0] + ::Tool::ModuleBuild => 0 : Build.PL does not exist
    [Path::IsDev=0] + ::META => 0 : META.json does not exist
    [Path::IsDev=0] + ::META => 1 : META.yml exists
    [Path::IsDev=0] + ::META => 1 : /home/kent/perl5/META.yml is a file
    [Path::IsDev=0] + ::META matched path /home/kent/perl5
    /home/kent/perl5

Whoops!.

    [Path::IsDev=0] + ::META => 1 : META.yml exists
    [Path::IsDev=0] + ::META => 1 : /home/kent/perl5/META.yml is a file

No wonder!

    rm /home/kent/perl5/META.yml

    PATH_ISDEV_DEBUG=1 \
        perl -Ilib -MPath::FindDev=find_dev \
        -E "say find_dev(q{/home/kent/perl5/perlbrew/perls/perl-5.19.3/lib/site_perl})"

    ...
    [Path::IsDev=0] Matching /home/kent/perl5
    ...
    [Path::IsDev=0] + ::TestDir => 0 : xt does not exist
    [Path::IsDev=0] + ::TestDir => 1 : t exists
    [Path::IsDev=0] + ::TestDir => 1 : /home/kent/perl5/t is a dir
    [Path::IsDev=0] + ::TestDir matched path /home/kent/perl5
    /home/kent/perl5

Double whoops!

    [Path::IsDev=0] + ::TestDir => 1 : t exists
    [Path::IsDev=0] + ::TestDir => 1 : /home/kent/perl5/t is a dir

And you could keep doing that until you rule out all the bad heuristics in your tree.

Or, you could use a negative heuristic.

    touch /home/kent/perl5/.path_isdev_ignore

    PATH_ISDEV_DEBUG=1 \
        perl -Ilib -MPath::FindDev=find_dev \
        -E "say find_dev(q{/home/kent/perl5/perlbrew/perls/perl-5.19.3/lib/site_perl})"
    ...
    [Path::IsDev=0] Matching /home/kent/perl5
    [Path::IsDev=0] - ::IsDev::IgnoreFile => 1 : .path_isdev_ignore exists
    [Path::IsDev=0] - ::IsDev::IgnoreFile => 1 : /home/kent/perl5/.path_isdev_ignore is a file
    [Path::IsDev=0] - ::IsDev::IgnoreFile excludes path /home/kent/perl5
    [Path::IsDev=0] no match found
    ...
    [Path::IsDev=0] Matching /
    ...
    [Path::IsDev=0] no match found

Success!

    [Path::IsDev=0] - ::IsDev::IgnoreFile => 1 : .path_isdev_ignore exists
    [Path::IsDev=0] - ::IsDev::IgnoreFile => 1 : /home/kent/perl5/.path_isdev_ignore is a file

=head1 HEURISTICS

=head2 Negative Heuristics bundled with this distribution

Just remember, a B<Negative> Heuristic B<excludes the path it is associated with>

=over 4

=item * L<< C<IsDev::IgnoreFile>|Path::IsDev::NegativeHeuristic::IsDev::IgnoreFile >> - C<.path_isdev_ignore>

=back

=head2 Positive Heuristics bundled with this distribution

=over 4

=item * L<< C<Changelog>|Path::IsDev::Heuristic::Changelog >> - Files matching C<Changes>, C<Changelog>, and similar, case
insensitive, extensions optional.

=item * L<< C<DevDirMarker>|Path::IsDev::Heuristic::DevDirMarker >> - explicit C<.devdir> file to indicate a project root.

=item * L<< C<META>|Path::IsDev::Heuristic::META >> - C<META.yml>/C<META.json>

=item * L<< C<MYMETA>|Path::IsDev::Heuristic::MYMETA >> - C<MYMETA.yml>/C<MYMETA.json>

=item * L<< C<Makefile>|Path::IsDev::Heuristic::Makefile >> - Any C<Makefile> format documented supported by GNU Make

=item * L<< C<TestDir>|Path::IsDev::Heuristic::TestDir >> - A directory called either C<t/> or C<xt/>

=item * L<< C<Tool::DZil>|Path::IsDev::Heuristic::Tool::DZil >> - A C<dist.ini> file

=item * L<< C<Tool::MakeMaker>|Path::IsDev::Heuristic::Tool::MakeMaker >> - A C<Makefile.PL> file

=item * L<< C<Tool::ModuleBuild>|Path::IsDev::Heuristic::Tool::ModuleBuild >> - A C<Build.PL> file

=item * L<< C<VCS::Git>|Path::IsDev::Heuristic::VCS::Git >> - A C<.git> directory

=back

=head1 HEURISTIC SETS

=head2 Heuristic Sets Bundled with this distribution

=over 4

=item * L<< C<Basic>|Path::IsDev::HeuristicSet::Basic >> - The basic heuristic set that contains most, if not all heuristics.

=back

=head1 ADVANCED USAGE

=head2 Custom Sets

C<Path::IsDev> has a system of "sets" of Heuristics, in order to allow for pluggable
and flexible heuristic types.

Though, for the vast majority of cases, this is not required.

    use Path::IsDev is_dev => { set => 'Basic' };
    use Path::IsDev is_dev => { set => 'SomeOtherSet' , -as => 'is_dev_other' };

=head2 Overriding the default set

If for whatever reason the C<Basic> set is insufficient, or if it false positives on your system for some reason,
the "default" set can be overridden.

    export PATH_ISDEV_DEFAULT_SET="SomeOtherSet"

    ...
    use Path::IsDev qw( is_dev );
    is_dev('/some/path') # uses SomeOtherSet

Though this will only take priority in the event the set is not specified during C<import>

If this poses a security concern for the user, then this security hole can be eliminated by declaring the set you want in code:

    export PATH_ISDEV_DEFAULT_SET="SomeOtherSet"

    ...
    use Path::IsDev  is_dev => { set => 'Basic' };
    is_dev('/some/path') # uses Basic, regardless of ENV

=head1 SECURITY

Its conceivable, than an evil user could construct an evil set, containing arbitrary and vulnerable code, and possibly stash that
evil set in a poorly secured privileged users @INC

And if they managed to achieve that, if they could poison the privileged users %ENV, they could trick the privileged user into
executing arbitrary code.

Though granted, if you can do either of those 2 things, you're probably security vulnerable anyway, and granted, if you could do
either of those 2 things you could do much more evil things by the following:

    export PERL5OPT="-MEvil::Module"

So with that in understanding, saying this modules default utility is "insecure" is mostly a bogus argument.

And to that effect, this module does nothing to "lock down" that mechanism, and this module encourages you
to B<NOT> force a set, unless you B<NEED> to, and strongly suggests that forcing a set for the purpose of security will achieve
no real improvement in security, while simultaneously reducing utility.

=head1 AUTHOR

Kent Fredric <kentnl@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2017 by Kent Fredric <kentfredric@gmail.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
